library/general/annotationrestriction/template.yaml

```diff
@@ -0,0 +1,33 @@
+apiVersion: templates.gatekeeper.sh/v1beta1
+kind: ConstraintTemplate
+metadata:
+  name: requiredannotations
+spec:
+  crd:
+    spec:
+      names:
+        kind: RequiredAnnotations
+        listKind: RequiredAnnotationsList
+        plural: requiredannotations
+        singular: requiredannotations
+      validation:
+        # Schema for the `parameters` field in the constraint
+        openAPIV3Schema:
+          properties:
+          #Name of the annotation you want to force the constraint on
+            annotationKey:
+              type: string
+          #Value of annotation you want to enforce
+            annotationValue:
+              type: string
+  targets:
+    - target: admission.k8s.gatekeeper.sh
+      rego: |
+        package requiredannotations
+
+        violation[{"msg": msg, "details": {}}] {
```

Comment:
We should also have a `violation[]` rule that checks whether `input.review.object.metadata.annotations[input.parameters.annotationKey]` is defined at all :

```
not input.review.object.metadata.annotations[input.parameters.annotationKey]
```

That way the user can require that the annotation exists

Reply:
I think that when we've deployed this constrainttemplate + constraint, it is already blocking the creation of ingresses with the annotation unset.

Based on that, I don't think the line you suggested is needed?

Reply:
Weird, it does work, but is inconsistent with the REPL.

@tsandall my OPA shell does this:

```
> not data.asdf == 2
undefined
```

which suggests that the rule shouldn't match if asdf doesn't exist (as undefined => no match). Is there a subtlety I'm missing or is this a genuine inconsistency?

@atheiman
+1 the line is not needed